### DISCUSSION TITLE
How is QUIC_MAX_PARTITION_COUNT set?

### DISCUSSION THREAD

**Discussion by jeremyong** [microsoft/msquic#1433] [Q&A] [answered]
Quick question, I see that `QUIC_MAX_PARTITION_COUNT` is defined in `quicdef.h` and that the docs specify this can be set dynamically. However, I don't see a corresponding setting in `QUIC_SETTING`s and it seems like the IO threads are spawned before the registration even occurs (as part of the library's initialization routine).

The primary use case for this would be if you need to do some heavier server work and don't want other threads thrashing cache memory on certain cores. It looks like in realtime mode, thread affinity is assigned from 0 counting upwards per partition. Another use case is specifically for client side use where currently all available cores immediately map to two threads on launch (even if you only have one connection intended for lighter usage).

**Comment by nibanks** [comment, marked as answer]
We have a external storage (i.e. registry) value to control it, but assuming that's not set, we just set it to the processor count (capped to `QUIC_MAX_PARTITION_COUNT`). See https://github.com/microsoft/msquic/blob/main/src/core/library.c#L304. A `QUIC_SETTING` from the app isn't going to do anything since we initialize that at the first call to `MsQuicOpen`.

**Reply by jeremyong** [reply]
https://github.com/jeremyong/msquic/commit/957da1bb8997410203948f77f3c530cea5a0b1b5

Here you can see the start of a small change to support a user-defined affinity bitmask and max worker count. The issue is that this code as is won't work because `CxPlatInitialize` happens in `MsQuicLibraryInitialize` well before the registration occurs.

**Reply by jeremyong** [reply]
Are there any plans to support changing it programmatically by something other than registry value?

**Reply by nibanks** [reply]
No we don't currently have explicit plans for it, but that doesn't mean we're against it necessarily.

Why doesn't your commit help? Because it only affects the worker pool, and not the datapath threads?

**Reply by jeremyong** [reply]
Well I noticed two problems. One is that the client connection crashes immediately because the worker that gets assigned to the connection is corrupted (haven't investigated why). The second problem is that on the server, when a listener is spawned, the user-supplied registration is ignored and the internal registration is used instead (which ends up spawning the max number of partitions based on cpu count).

**Reply by jeremyong** [reply]
The other problem with the datapath threads spawning at max rate exists also (those threads are spawned before the user registration is known as you point out in your top level response)

**Reply by nibanks** [reply]
For the datapath threads, I argue they shouldn't be a problem. They're only used if RSS delivers to that core. You can always configure RSS not to use those cores then.

**Reply by jeremyong** [reply]
OK I'll look into it when I have time thanks for the responses.

**Reply by nibanks** [reply]
Sure. Also, in Windows kernel mode, there are no threads created, so you can't control where you get receives from, beyond configuring RSS. I'd prefer to keep that logic consistent across the different abstraction layers.

**Reply by jeremyong** [reply]
Makes sense, it seems like the easiest change to make is to have `quicdef.h` set the defines for each param only if they aren't already defined. That way, the user can just override a setting as needed system-wide by defining the macro before including the quic header. Would you be interested in that change?

edit: Oh nevermind these macros get compiled into source haha

**Reply by nibanks** [reply]
I was just about to say... :) quicdef.h is internal to the library and not exposed to apps.

**Reply by jeremyong** [reply]
Yea whoops!
